### PR TITLE
Send pip status events through to SSE

### PR DESCRIPTION
**Pull request by joeyballentine**
This PR sets up better dep installation progress reporting. It also starts the beginning of the pip output parsing that will hopefully be used in the future to give download progress.

### CODE REVIEW

File: backend/src/dependencies/store.py [hunk 1]
```diff
@@ -63,13 +60,140 @@ def install_dependencies(dependencies: List[DependencyInfo]):
...
+        version = dep_info["version"]
+        installed_packages[package_name] = version
+
+
+def install_dependencies_sync(
+    dependencies: List[DependencyInfo],
+):
+    dependencies_to_install = get_deps_to_install(dependencies)
+    if len(dependencies_to_install) > 0:
+        install_dependencies_sync_impl(dependencies_to_install)
```
Comment (RunDevelopment) [inline backend/src/dependencies/store.py:92]: Not a fan of those `_impl`. I believe python has module-private functions (`_name` I believe), so please use those when you need private things.

But do we need to split the this into 2 at all? Why not just:

```py
def install_dependencies_sync(
    dependencies: List[DependencyInfo],
):
    dependencies= get_deps_to_install(dependencies)
    if len(dependencies) == 0:
        return
    # impl
```

Same for the async version.

**Review by RunDevelopment — COMMENTED** [review]
The dependency counting seems unreliable right now. If I understand the code correctly, we count the packages we want to install + their (transitive) dependencies. So we will in all likelihood always overshoot `dep_length`.

Maybe we could so a little trick. How about make 2 separate counters: one for the packages and we actually want to install and one for the transitive dependencies (if any). This means that we can calculate progress based on the number of packages we requested. I would propose the following formula for the progress:

```
# approach 1 for tansitive deps
transitive_progress = 1 - 1 / (2**transitive_deps_counter)
progress = (deps_counter + transitive_progress) / (deps_count + 1)
```

Obviously, this progress has to be moved into the range you were using, but I'll leave that to you.

**Comment by joeyballentine** [comment]
I think you're misunderstanding the code a bit. Right now it assumes just the number of requested dependencies, and then counts every dependency that gets installed. But, it clamps it to never actually overshoot. This means that it might sit at 100% of the subprogress area for a while while it installs the downstream dependencies, which in practice is only a few extras. I personally don't see this as a big deal

File: backend/src/dependencies/store.py [hunk 2]
```diff
@@ -63,13 +60,140 @@ def install_dependencies(dependencies: List[DependencyInfo]):
...
+
+
+def log_impl(message: str):
+    if logger is not None:
+        logger.info(message)
+    else:
+        print(message, flush=True)
```
[the remark this answers is not in the document]
Comment (joeyballentine) [reply backend/src/dependencies/store.py:99]: Whoops, i meant to get rid of this. This was just something i did for debugging purposes

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**